src/client/components/common/Combobox.tsx

```diff
@@ -0,0 +1,358 @@
+    selectedItem: FieldOption
+    topDropdownInset: number
+    getItemProps: (options: GetItemPropsOptions<ComboboxItem>) => unknown
+  }
+}
+
+/**
+ * Renders a single dropdown item.
+ *
+ * Used by react-window's `FixedSizeList` to efficiently render
+ * large lists. This helps reduce the performance impact of
+ * Downshift's event re-renders.
+ *
+ * @param props Props passed in by the `FixedSizeList` pure component.
+ *
+ * @returns A list item to be displayed in the dropdown
+ */
+const ItemRenderer: FC<ItemRendererProps> = (props) => {
+  const { items, getItemProps, highlightedIndex, topDropdownInset } = props.data
+  const item = items[props.index]
+
+  return (
+    <>
+      <ListItem
+        {...getItemProps({
+          style: {
+            ...props.style,
+            // purely for adding padding insets to FixedStyleList
+            // see https://codesandbox.io/s/react-window-list-padding-dg0pq
+            top: `${parseFloat(`${props.style.top}`) + topDropdownInset}px`,
+          },
+          item,
+          index: props.index,
+        })}
+        bg={highlightedIndex === props.index ? 'neutral.50' : 'none'}
+        py={'12px'}
+        px={'20px'}
+      >
+        <HStack spacing={4}>
+          <Text isTruncated>{item.label}</Text>
+        </HStack>
+      </ListItem>
+    </>
+  )
+}
+
+interface ComboboxProps extends Omit<InputProps, 'onChange' | 'label'> {
+  items: ComboboxItem[]
+  label: string
+  dropdownOptions?: {
+    height?: number
+    itemHeight?: number
+    inset?: number
+    dropupThreshold?: number
+  }
+  searchOptions?: MatchSorterOptions<ComboboxItem>
+  inputOptions?: {
+    forwardRef?: RefObject<HTMLInputElement>
+    useClearButton?: boolean
+  }
+  onChange: (value: unknown) => void
+}
+
+/**
+ * Renders a combobox dropdown.
+ *
+ * @param items The items to render within the dropdown.
+ * @param dropdownOptions Configuration options for rendering the dropdown.
+ * @param searchOptions Configuration options for `matchSorter` to use when searching
+ * through the dropdown based on the user input.
+ * @param inputOptions Configuration options for rendering the combobox input.
+ * @returns A combobox element
+ */
+export const Combobox: FC<ComboboxProps> = ({
+  items,
+  label,
+  dropdownOptions,
+  searchOptions,
+  inputOptions,
+  onChange,
+  ...props
+}) => {
+  const dropdownHeight = dropdownOptions?.height || 200
+  const itemHeight = dropdownOptions?.itemHeight || 40
+  const dropdownInset = dropdownOptions?.inset || 0
+
+  // Avoid using refs for dropdown direction as it's only used
+  let dropdownDir = DropdownDirection.down
+
+  // Note that inputRef is a merged ref as we may not want to
+  // always provide a forward ref for the input
+  const fallbackInputRef = useRef<HTMLInputElement>(null)
+  const inputRef = inputOptions?.forwardRef || fallbackInputRef
+
+  const dropdownList = useRef<FixedSizeList & HTMLElement>(null)
+  const searchResults = useRef<ComboboxItem[]>(items)
```

Comment:
Is there as specific reason to use `useRef` instead of `useState`? 

Generally, `useRef` should be used for things that persists throughout the entire lifecycle of a component and `useState` is used for things that causes a re-render. Given that when the search results changes, a re-render would be necessary, might be more appropriate to use `useState`.